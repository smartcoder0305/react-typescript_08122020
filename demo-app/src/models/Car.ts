export type Car = {
  id: number;
  make: string;
  model: string;
  year: number;
  color: string;
  price: number;
};
