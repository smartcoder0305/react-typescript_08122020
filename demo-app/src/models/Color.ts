

export type Color = {
  id: number,
  name: string,
  hexcode: string,
};