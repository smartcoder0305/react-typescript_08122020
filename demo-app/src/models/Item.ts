export interface Item {
  id?: number;
}