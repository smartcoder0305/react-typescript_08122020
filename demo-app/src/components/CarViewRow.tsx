import React, { FC } from 'react';

import { Car } from '../models/Car';

export type CarViewRowProps = {
  car: Car,
};

export const CarViewRow: FC<CarViewRowProps> = ({ car }) => {

  return (
    <tr>
      <td>{car.id}</td>
      <td>{car.make}</td>
      <td>{car.model}</td>
      <td>{car.year}</td>
      <td>{car.color}</td>
      <td>{car.price}</td>
    </tr>
  );

};