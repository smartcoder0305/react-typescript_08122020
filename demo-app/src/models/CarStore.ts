import { Car } from '../models/Car';
import { CarTableFilterSettings } from '../components/CarTableFilterForm';
import { CarTableSortSettings } from '../components/CarTable';

export type CarStore = {
  cars: Car[],
  editCarId: number,
  carTableSortSettings: CarTableSortSettings,
  confirmDeleteCarId: number,

  setEditCarId: (carId: number) => void,
  confirmDeleteCar: (carId: number) => void,
  deleteCar: (carId: number) => void,
  getCarDetailsById: (carId: number) => string,

  cancelCar: () => void,
  dismissConfirmDeleteCarModal: () => void,
  saveCar: (car: Car) => void,
  addCar: (car: Car) => void,

  setCarTableFilterSettings: (filterSettings: CarTableFilterSettings) => void,
  setCarTableSortSettings: (sortSettings: CarTableSortSettings) => void,
};