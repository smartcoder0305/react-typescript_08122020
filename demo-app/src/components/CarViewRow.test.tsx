import * as React from 'react';
import {
  render, mount, shallow,
  ReactWrapper, ShallowWrapper,
} from 'enzyme';

import { Car } from '../models/Car';
import { CarViewRow } from './CarViewRow';

describe('<CarViewRow /> Enzyme Static HTML', () => {

  let car: Car;

  beforeEach(() => {
    car = {
      id: 1,
      make: 'Ford',
      model: 'F-150',
      year: 1980,
      color: 'red',
      price: 42000,
    };
  });

  test('<CarViewRow /> renders', () => {
    const component = JSON.stringify(
      render(
        <table>
          <tbody>
            <CarViewRow
              car={car}
              onDeleteCar={() => null}
              onEditCar={() => null}
            />
          </tbody>
        </table>,
      ).html(),
    );

    expect(component).toMatchSnapshot();
  });
});

describe('<CarViewRow /> Enzyme Mock DOM', () => {

  const eventHandlers = {
    deleteCar: () => null,
    editCar: () => null,
  };

  let car: Car;
  let deleteCarSpy: jest.SpyInstance;
  let editCarSpy: jest.SpyInstance;
  let component: ReactWrapper;

  beforeEach(() => {
    car = {
      id: 1,
      make: 'Ford',
      model: 'F-150',
      year: 1980,
      color: 'red',
      price: 42000,
    };

    deleteCarSpy = jest.spyOn(eventHandlers, 'deleteCar');
    editCarSpy = jest.spyOn(eventHandlers, 'editCar');

    component = mount(
      <table>
        <tbody>
          <CarViewRow
            car={car}
            onDeleteCar={eventHandlers.deleteCar}
            onEditCar={eventHandlers.editCar}
          />
        </tbody>
      </table>,
    ).find(CarViewRow);
  });

  test('<CarViewRow /> renders', () => {
    const columns = ['id', 'make', 'model', 'year', 'color', 'price'];

    component
      .find('td')
      .slice(0, 6)
      .forEach((node, index) => {
        const carField = String(car[columns[index]]);
        expect(node.text()).toBe(carField);
      });
  });

  test('<CarViewRow /> delete car button', () => {
    component
      .find('button')
      .first()
      .simulate('click');
    component
      .find('button')
      .last()
      .simulate('click');

    expect(deleteCarSpy).toHaveBeenCalledWith(car.id);
    expect(editCarSpy).toHaveBeenCalledWith(car.id);
  });
});

describe('<CarViewRow /> Shallow with Enzyme', () => {
  const eventHandlers = {
    deleteCar: () => null,
    editCar: () => null,
  };

  let car: Car;
  let deleteCarSpy: jest.SpyInstance;
  let editCarSpy: jest.SpyInstance;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    car = {
      id: 1,
      make: 'Ford',
      model: 'F-150',
      year: 1980,
      color: 'red',
      price: 42000,
    };

    deleteCarSpy = jest.spyOn(eventHandlers, 'deleteCar');
    editCarSpy = jest.spyOn(eventHandlers, 'editCar');

    wrapper = shallow(
      <CarViewRow
        car={car}
        onDeleteCar={eventHandlers.deleteCar}
        onEditCar={eventHandlers.editCar}
      />,
    );
  });

  test('<CarViewRow /> renders', () => {
    const columns = ['id', 'make', 'model', 'year', 'color', 'price'];

    wrapper
      .find('td')
      .slice(0, 6)
      .forEach((node, index) => {
        const carField = String(car[columns[index]]);
        expect(node.text()).toBe(carField);
      });
  });

  test('<CarViewRow /> buttons', () => {
    wrapper
      .find('button')
      .first()
      .simulate('click');
    wrapper
      .find('button')
      .last()
      .simulate('click');

    expect(deleteCarSpy).toHaveBeenCalledWith(car.id);
    expect(editCarSpy).toHaveBeenCalledWith(car.id);
  });
});
