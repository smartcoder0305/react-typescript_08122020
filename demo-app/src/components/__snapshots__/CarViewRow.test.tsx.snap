// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarViewRow /> Enzyme Static HTML <CarViewRow /> renders 1`] = `"\\"<tbody><tr><td>1</td><td>Ford</td><td>F-150</td><td>1980</td><td>red</td><td>42000</td><td><button type=\\\\\\"button\\\\\\">Edit</button><button type=\\\\\\"button\\\\\\">Delete</button></td></tr></tbody>\\""`;
